### PR TITLE
Metrics - Cloudwatch v2

### PR DESCRIPTION
**Pull request by nelz9999**
Per the conversation in the Slack channel: https://gophers.slack.com/archives/C04S3T99A/p1519410405000323

I have created the `metrics/cloudwatch2` package.

This package targets using the V2 SDK (https://github.com/aws/aws-sdk-go-v2). It also opts for using the CloudWatch native client-side aggregation into "StatisticsSet" -
 https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_StatisticSet.html. (Whereas the existing `metrics/cloudwatch` package does addition lossy aggregations and just emits a single "Value".) Also, as per CloudWatch idiom, Counters / Histograms / Gauges are all sent to CloudWatch in exactly the same way.

(I mostly cribbed from the existing `cloudwatch` package, simplified the work it does, and touched up around the edges to adjust to the delta in API between v1 and v2 versions of the AWS SDK.)

Since Histograms/Gauges all reduce down to the exact same behavior as Counters, I created the `convert` package to easily swap in Histogram/Gauge API semantics. I made it its own package in case any other metrics systems have a need for similar generification... Though if this is an overshoot, I could dumb that down and just isolate it into the `cloudwatch2` package.

### CODE REVIEW

File: metrics/cloudwatch2/cloudwatch2.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		if lim > maxConcurrentRequests {
+			lim = maxConcurrentRequests
+		}
+		batch, datums = datums[:lim], datums[lim:]
+		batches = append(batches, batch)
+	}
+
+	var errors = make(chan error, len(batches))
+	for _, batch := range batches {
+		go func(batch []cloudwatch.MetricDatum) {
```
Comment (groob) [inline metrics/cloudwatch2/cloudwatch2.go]: the goroutines could run as an [errorgroup](https://godoc.org/golang.org/x/sync/errgroup)

File: metrics/cloudwatch2/cloudwatch2.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	namespace             string
+	svc                   cloudwatchiface.CloudWatchAPI
+	counters              *lv.Space
+	logger                log.Logger
+	numConcurrentRequests int
+}
+
+type option func(*CloudWatch)
+
+func (cw *CloudWatch) apply(opt option) {
```
[the remark this answers is not in the document]
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Ahh, yup. That's what I get for doin' a copypasta.

File: metrics/cloudwatch2/cloudwatch2.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+// Callers must ensure that regular calls to Send are performed, either
+// manually or with one of the helper methods.
+func New(namespace string, svc cloudwatchiface.CloudWatchAPI, options ...option) *CloudWatch {
+	cw := &CloudWatch{
+		sem:                   nil, // set below
+		namespace:             namespace,
+		svc:                   svc,
+		counters:              lv.NewSpace(),
+		numConcurrentRequests: 10,
+		logger:                log.NewLogfmtLogger(os.Stderr),
```
[the remark this answers is not in the document]
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Fixed.

File: metrics/cloudwatch2/cloudwatch2.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+	}
+}
+
+// New returns a CloudWatch object that may be used to create metrics.
+// Namespace is applied to all created metrics and maps to the CloudWatch namespace.
+// Callers must ensure that regular calls to Send are performed, either
+// manually or with one of the helper methods.
+func New(namespace string, svc cloudwatchiface.CloudWatchAPI, options ...option) *CloudWatch {
+	cw := &CloudWatch{
+		sem:                   nil, // set below
```
[the remark this answers is not in the document]
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Will do!

File: metrics/cloudwatch2/cloudwatch2.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		if err := <-errors; err != nil && firstErr != nil {
+			firstErr = err
+		}
+	}
+
+	return firstErr
+}
+
+var zero = float64(0.0)
+var zeros = cloudwatch.StatisticSet{
```
[the remark this answers is not in the document]
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: My intent was to avoid paying the "construction costs" every time a zero set is needed.

File: metrics/cloudwatch2/cloudwatch2.go (hunk 1 above)
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Oooh, cool! I hadn't yet added `errorgroup` to my toolbelt. Added!

File: metrics/cloudwatch2/cloudwatch2.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: ... Should I do something different to make that apparent?

[the remark this answers is not in the document]
Comment (peterbourgon) [reply metrics/cloudwatch2/cloudwatch2.go]: Maybe just a light comment could work.

[the remark this answers is not in the document]
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Sure thing!

**Comment by nelz9999** [comment]
@peterbourgon How are you feeling about this PR?

File: metrics/cloudwatch2/cloudwatch2.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+	mtx                   sync.RWMutex
+	sem                   chan struct{}
+	namespace             string
+	svc                   cloudwatchiface.CloudWatchAPI
+	counters              *lv.Space
+	logger                log.Logger
+	numConcurrentRequests int
+}
+
+type option func(*CloudWatch)
```
Comment (peterbourgon) [inline metrics/cloudwatch2/cloudwatch2.go]: Since this is returned by exported functions, it's nicer to also make it exported.

File: metrics/cloudwatch2/cloudwatch2.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+// WithLogger sets the Logger that will recieve error messages generated
+// during the WriteLoop
+func WithLogger(logger log.Logger) option {
```
Comment (peterbourgon) [inline metrics/cloudwatch2/cloudwatch2.go]: I find it's nice if functional options describe the default behavior if the option _isn't_ invoked. In this case, it might be nice to add: "By default, no logger is used."

File: metrics/cloudwatch2/cloudwatch2.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+	return func(cw *CloudWatch) {
+		cw.logger = logger
+	}
+}
+
+// WithConcurrentRequests sets the upper limit on how many
+// cloudwatch.PutMetricDataRequest may be under way at any
+// given time. If n is greater than 20, 20 is used.
+func WithConcurrentRequests(n int) option {
```
Comment (peterbourgon) [inline metrics/cloudwatch2/cloudwatch2.go]: Similarly: "By default, 10 concurrent requests are used."

File: metrics/cloudwatch2/cloudwatch2.go [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+package cloudwatch2
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch/cloudwatchiface"
```
Comment (peterbourgon) [inline metrics/cloudwatch2/cloudwatch2.go]: package cloudwatchiface 🤦‍♂️

File: metrics/cloudwatch2/cloudwatch2.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+		Maximum:     &max,
+		Minimum:     &min,
+		Sum:         &sum,
+		SampleCount: &count,
+	}
+}
+
+func makeDimensions(labelValues ...string) []cloudwatch.Dimension {
+	dimensions := make([]cloudwatch.Dimension, len(labelValues)/2)
+	for i, j := 0, 0; i < len(labelValues); i, j = i+2, j+1 {
```
Comment (peterbourgon) [inline metrics/cloudwatch2/cloudwatch2.go]: Ha! Wild. Nice.

File: metrics/convert/convert.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
+// Package convert provides a way to use Counters, Histograms, or Gauges
+// as one of the other types
+package convert
```
Comment (peterbourgon) [inline metrics/convert/convert.go]: It feels like this is tailor-made to cloudwatch2, and I'm a little bit scared that folks will see it and think it's something they should be doing in their own code. Would you be OK with moving this underneath the internal/ dir?

File: metrics/cloudwatch2/cloudwatch2_test.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+				t.Errorf("expected [%f]: %f\n", tc.xSum, *s.Sum)
+			}
+			if tc.xCt != *s.SampleCount {
+				t.Errorf("expected [%f]: %f\n", tc.xCt, *s.SampleCount)
+			}
+		})
+	}
+}
+
+type mockCloudWatch struct {
```
Comment (peterbourgon) [inline metrics/cloudwatch2/cloudwatch2_test.go]: Great.

File: metrics/convert/convert_test.go [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/go-kit/kit/metrics/teststat"
+)
+
+func TestCounterHistogramConversion(t *testing.T) {
+	name := "my_counter"
+	c := generic.NewCounter(name)
+	h := NewCounterAsHistogram(c)
+	top := NewHistogramAsCounter(h).With("label", "counter").(histogramCounter)
+	mid := top.h.(counterHistogram)
+	low := mid.c.(*generic.Counter)
```
Comment (peterbourgon) [inline metrics/convert/convert_test.go]: Hehe, nice!

File: metrics/cloudwatch2/cloudwatch2.go (hunk 6 above)
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Will do.

File: metrics/cloudwatch2/cloudwatch2.go (hunk 7 above)
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Fixing...

File: metrics/cloudwatch2/cloudwatch2.go (hunk 8 above)
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Adding...

File: metrics/cloudwatch2/cloudwatch2.go (hunk 9 above)
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: Yah. It's supposed to be in support of mocking out their service... But it feels so icky.

File: metrics/cloudwatch2/cloudwatch2.go (hunk 10 above)
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2.go]: I can't take credit for this... I lifted this directly from the existing `cloudwatch` package.

File: metrics/convert/convert.go (hunk 11 above)
Comment (nelz9999) [reply metrics/convert/convert.go]: Will do!

File: metrics/cloudwatch2/cloudwatch2_test.go (hunk 12 above)
Comment (nelz9999) [reply metrics/cloudwatch2/cloudwatch2_test.go]: :blush:

File: metrics/convert/convert_test.go (hunk 13 above)
Comment (nelz9999) [reply metrics/convert/convert_test.go]: 🙇

**Comment by nelz9999** [comment]
@peterbourgon Back atcha! 😄

**Comment by peterbourgon** [comment]
Nice! Thanks!
